### PR TITLE
LPS-187797 Add setting for prefer reduced motion

### PR DESCRIPTION
**Pull request by matuzalemsteles**
Ticket: https://liferay.atlassian.net/browse/LPS-187797

I made some changes here so that any non-Clay application that implements JS animation can consume the accessibility menu information, I changed the accessibility menu settings state to use the global state.

```jsx
import {accessibilityMenuAtom} from '@liferay/accessibility-menu-web';
import {useLiferayState} from '@liferay/frontend-js-react-web';

const [accessibilityMenu] = useLiferayState(accessibilityMenuAtom);
```

I added the Clay Provider configuration in the `render.js` which is used by most of the DXP applications, so we can reach all the applications that are React.js and the Clay components, basically we are consuming the global state of Liferay with the atom which is public when consuming the `@liferay/accessibility-menu-web` package and we look for the reduced motion setting to configure the provider. Probably not all applications are consuming `render.js` so we will probably have to replicate this same behavior in these applications or just use `render.js` to avoid this whole configuration boilerplate.

I'm having some compilation issues with the `frontend-js-react-web` module not finding the `@liferay/accessibility-menu-web` module which doesn't make much sense, so submitting this to check if it's just a problem on my machine or configuration error. We will see!

### CODE REVIEW

File: modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx
Comment (matuzalemsteles) [inline modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx:135]: This could probably be a constant exported by the `@liferay/accessibility-menu-web` module.

**Comment by matuzalemsteles** [comment]
Looks like it's a configuration error, I'll check that tomorrow when it starts working.

**Review by ethib137 — CHANGES_REQUESTED** [review]
Looks good, just a few suggestions. Thanks @matuzalemsteles

File: modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.js
Comment (matuzalemsteles) [reply modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.js:38]: Well, maybe so, but I think it's a bit exaggerated at this point if you're going to think about performance, it's a very small list and using both approaches we won't have any perceptible improvement in that regard. I think it would be more a matter of convenience if that were the case in adopting this.

In that case the `settingKeys` could be a `useMemo` on the first render of the component or just calling it directly on render would also have no negative performance effects.

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
I made the changes, got the commits from Bryce which solves the configuration issues and also converted the module to TypeScript for easier implementation.

**Comment by matuzalemsteles** [comment]
ci:test:sf

File: modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.js
Comment (ethib137) [reply modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.js:60]: Yeah, I wasn't thinking about it as a performance improvement. I was thinking more that since we are adding this to LiferayState, it ends up being our api for accessing settings. It's a much simpler api to have direct access to the setting based on it's key rather than having to loop through the settings to find the setting you're looking for.

File: modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.tsx
Comment (ethib137) [inline modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.tsx:125]: What's the reason for adding this to the dependencies of this hook, instead of passing a function to `setSettings` that can access the current value of settings directly? Wouldn't this result in infinite re-renders since `settings` would change every time we call `setSettings`?

File: modules/apps/accessibility/accessibility-menu-web/types/src/main/resources/META-INF/resources/js/index.d.ts
Comment (ethib137) [inline modules/apps/accessibility/accessibility-menu-web/types/src/main/resources/META-INF/resources/js/index.d.ts:20]: The constants have been renamed for consistency, so rebasing and renaming here to match is needed. https://github.com/liferay/liferay-portal/blob/master/modules/apps/accessibility/accessibility-menu-web/src/main/java/com/liferay/accessibility/menu/web/internal/constants/AccessibilitySettingConstants.java

File: modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.tsx
Comment (matuzalemsteles) [reply modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.tsx:125]: @ethib137 I thought this was working but the set state of `useLiferayState` doesn't support function so it wasn't working before. But in this case it won't cause an infinite loop because this method is not used as a dependency on `useEffect` for example.

**Comment by matuzalemsteles** [comment]
ci:test:sf

File: modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.js
Comment (matuzalemsteles) [reply modules/apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.js:59]: Got it, I think if we're just thinking about simplifying the public API it makes sense, performance we won't see any optimization effect even if it's looping to find or update.

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
ci:stop

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
Hey @matuzalemsteles, do you intend to push more commits to it or I may analyze the failures?

**Comment by matuzalemsteles** [comment]
Hi @eltonccdantas It looks like it's still failing a CI test, I think it's due to the change we made to the module's configuration. I'll look into that now.

**Comment by matuzalemsteles** [comment]
Hey @bryceosterhaus Could you help here to understand this error in the `document-library-opener-onedrive-web` module, it seems that due to some dependencies it is breaking in this module. The error also looks quite confusing and it doesn't really seem to be the error, maybe it's an AMD and ESM issue here?

Fail on test:

```
 FAIL  test/js/DocumentLibraryOpener.es.js
  ● Test suite failed to run

    TypeError: Cannot read properties of undefined (reading 'openModal')

      143 | export {default as selectFolder} from './liferay/util/select_folder';
      144 | export {default as sub} from './liferay/util/sub';
    > 145 | export {default as toggleBoxes} from './liferay/util/toggle_boxes';
          |                   ^
      146 | export {default as toggleControls} from './liferay/util/toggle_controls';
      147 | export {default as toggleDisabled} from './liferay/util/toggle_disabled';
      148 | export {default as toggleRadio} from './liferay/util/toggle_radio';

      at Object.get [as openModal] (../../../../apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/index.es.js:145:19)
      at test/js/DocumentLibraryOpener.es.js:14:237
          at Array.forEach (<anonymous>)
      at _objectSpread (test/js/DocumentLibraryOpener.es.js:14:177)
      at Object.user:/Users/matuzalemteles/LiferayWorkspace/liferay-portal/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/index.es.js: (test/js/DocumentLibraryOpener.es.js:16:30)
      at Object.<anonymous> (../../../../apps/cookies/cookies-banner-web/src/main/resources/META-INF/resources/cookies_banner/js/CookiesBanner.js:15:1)
      at Object.<anonymous> (../../../../apps/cookies/cookies-banner-web/src/main/resources/META-INF/resources/js/index.js:17:1)
      at Object.<anonymous> (../../../../apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/AccessibilityMenu.tsx:18:1)
      at Object.<anonymous> (../../../../apps/accessibility/accessibility-menu-web/src/main/resources/META-INF/resources/js/index.ts:15:1)
      at Object.<anonymous> (../../../../apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx:17:1)
      at Object.<anonymous> (../../../../apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/index.ts:19:1)
      at Object.<anonymous> (../../../../apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js:18:1)
      at Object.<anonymous> (../../../../apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/index.es.js:66:1)
      at Object.user:/Users/matuzalemteles/LiferayWorkspace/liferay-portal/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/index.es.js: (test/js/DocumentLibraryOpener.es.js:16:30)
      at Object.<anonymous> (src/main/resources/META-INF/resources/js/DocumentLibraryOpener.es.js:13:1)
      at Object.<anonymous> (test/js/DocumentLibraryOpener.es.js:12:1)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

**Comment by ethib137** [comment]
Hey @matuzalemsteles , Bryce will be on PTO until next week, so maybe in order to try and get this finished this week we could remove Bryce’s refactor as well as any Typescript. Without Typescript I think we no longer run into the apparent circular dependency. We can then create a separate technical task to add Typescript and Bryce’s refactor. Do you think that would work? Could you try that approach?

**Comment by matuzalemsteles** [comment]
Hi @ethib137 Thanks for letting me know, I'll try that but I don't think it's related to TypeScript, it seems that the circular dependency is due to the cookies-banner-web dependency which depends on the frontend-js-web and the frontend-js-web ends up depending on it because it uses `render` from frontend-js-react-web.

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:reevaluate:1267401_10208

**Comment by eltonccdantas** [comment]
Hey @matuzalemsteles, I saw that `js-unit-jdk8`  is marked as failed and contains the error message below:

```
console.error ../../../../node_modules/warning/warning.js:34
     [exec]       Warning: Button Accessibility: Component has only the Icon declared. Define an `aria-label` or `aria-labelledby` attribute that labels the interactive button that screen readers can read. The `title` attribute is optional but consult your design team.
     [exec]     console.error ../../../../node_modules/warning/warning.js:34
```

**Comment by matuzalemsteles** [comment]
@eltonccdantas this error is a common error in DXP and works only as a warning it doesn't make the test fail and I think it's not related to our change.

**Comment by eltonccdantas** [comment]
ci:test:stable

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
Now the problem is fixed, it was not related to TypeScript but the circular dependency to solve it I had to move the `CONSTANTS` and the `atom` from the global state of the accessibility menu to the `frontend-js-react-web` to avoid the circular dependency, unfortunately I would not be able to avoid this without making this change.

@john-co could you check if these failures are related to that PR?

**Comment by john-co** [comment]
Just started reviewing :)

**Comment by john-co** [comment]
Related:
- [CommerceCompare#CanHandleDeclinedCookiesUsingProductComparison](https://testray.liferay.com/home/-/testray/case_results/2366862325)
- [CookiesBanner#CookiesBannerAcceptAllSelected](https://testray.liferay.com/home/-/testray/case_results/2366863495)
- [CookiesConsentPanel#CookiesConsentPanelCouldSaveCookieSettingsViaPortletAtSiteLevel](https://testray.liferay.com/home/-/testray/case_results/2366863486)
- [CookiesConsentPanel#CookiesConsentPanelCouldSaveCookieSettingsWhenItsOnASitePage](https://testray.liferay.com/home/-/testray/case_results/2366863377)

**Comment by john-co** [comment]
Looks like the cookie banner doesn't show up after saving the cookie configuration

**[Actual]**

image

**[Expected]**

Selection_522

**Comment by matuzalemsteles** [comment]
Hmm that's weird, we didn't make any changes to that module maybe it's the change to ESM, I'll revert that since we don't need it anymore.

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
I decided to run the relevant suite again because of the aborted tests...

image

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:commerce

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
When I ran the tests locally, I found something that might be related

1. [com.liferay.portal.vulcan.internal.test.ReverseProxyTest ](https://gist.github.com/eltonccdantas/8f9cccedb60f1c30bbb7cfbf58554149#:~:text=com.liferay.portal.vulcan.internal.test.ReverseProxyTest%20%3E%20testActionsHrefValues%20FAILED)
2. [com.liferay.external.data.source.test.controller.test.ExternalDataSourceControllerParentTest](https://gist.github.com/eltonccdantas/39f08e552cabb986367ea868028e1a8f#:~:text=Suppressed%3A%20java.lang.AssertionError%3A%20Missing%20upgrade%20process%20created%20record)
3. [com.liferay.external.data.source.test.controller.test.ExternalDataSourceControllerTest](https://gist.github.com/eltonccdantas/0b445f5ab2856930a105874aeaa22485#:~:text=Suppressed%3A%20java.lang,Assert.java%3A89))
4. [com.liferay.external.data.source.test.controller.test.ExternalDataSourceProviderTest](https://gist.github.com/eltonccdantas/b2767e52197890c03465c902355c8974#:~:text=com.liferay.external.data.source.test.controller.test.ExternalDataSourceProviderTest%20%3E%20testExternalDataSourceTests%20FAILED)


Integration tests are still a bit challenging for me, do you think these failures are related to your changes?

**Comment by matuzalemsteles** [comment]
@eltonccdantas I think we can send it forward. cc @ethib137

File: modules/apps/frontend-js/frontend-js-react-web/types/src/main/resources/META-INF/resources/js/accessibilityMenu.d.ts
Comment (ethib137) [inline modules/apps/frontend-js/frontend-js-react-web/types/src/main/resources/META-INF/resources/js/accessibilityMenu.d.ts:1]: @matuzalemsteles this feels weird to add this to frontend-js-react-web, right? Is there a precedence for this type of inclusion?

File: modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js
Comment (ethib137) [inline modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: Hey @matuzalemsteles Based on [Bryce's comment ](https://github.com/liferay-peds/liferay-portal/pull/123#issue-1769018083)won't removing this cause it to be loaded as an AMD module instead of as an ESM module, which will make it unavailable to frontend-js-react-web. As a result this wouldn't work, right?

File: modules/apps/frontend-js/frontend-js-react-web/types/src/main/resources/META-INF/resources/js/accessibilityMenu.d.ts
Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-react-web/types/src/main/resources/META-INF/resources/js/accessibilityMenu.d.ts:1]: @ethib137 yeah but unfortunately that was the only way i found to fix the circular dependency problem, i added a comment about it [#122 (comment)](https://github.com/liferay-peds/liferay-portal/pull/122#issuecomment-1615225431), I think the amount of comment hid it.

Basically, the error it was causing in the tests is related to the circular dependency problem due to the `document-library-opener-onedrive-web` module -> `frontend-js-web` -> `frontend-js-react-web` -> `accessibility-menu-web` -> `cookies-banner-web` -> `frontend-js-web`. Also the cookies-banner-web module was [causing some strange test failures](https://github.com/liferay-peds/liferay-portal/pull/122#issuecomment-1615263602) due to turning the module to ESM.

File: modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js
Comment (matuzalemsteles) [reply modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: This is somewhat related to my previous comment, due to the changes to avoid the circular dependency now the `accessibility-menu-web` module no longer needs to be ESM the same for cookies because now we are no longer exposing the constants and the atom. I had to move to the `frontend-js-react-web` module to avoid this.

https://github.com/liferay-peds/liferay-portal/pull/122#issuecomment-1615225431

Comment (ethib137) [reply modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: Okay, I see how these are related. I don't think this is the way we want to fix the circular dependency since it actually ends up putting code that should be in the accessibility menu module in frontend-js-react-web. 

@matuzalemsteles Wouldn't another way of handling the circular dependency issue be to create a new module that exports our atom that both accessibility-menu-web and frontend-js-react-web could depend on? It could be something like accessibility-settings-state-web.

Comment (matuzalemsteles) [reply modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: > Wouldn't another way of handling the circular dependency issue be to create a new module that exports our atom that both accessibility-menu-web and frontend-js-react-web could depend on? It could be something like accessibility-settings-state-web.

That would be another solution as well, if that's something we want we can do that.

Comment (ethib137) [reply modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: Yeah, it's not ideal. But in general it seems Liferay is happier with having more small modules rather than adding unrelated code to another module. For instance what I want to avoid is in the future having to make a change to frontend-js-react-web when I'm adding a new accessibility menu setting.

If you're unsure about this, I'm happy to wait for Bryce to get back next week and we can ask him, or maybe someone else in frontend-infra could weigh in on this?

Comment (matuzalemsteles) [reply modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: I think we could go with this for now and create a ticket to make a better decision on this, whether we create a new module or move this elsewhere with help from Bryce. What do you think? I wanted to unblock this pull request as soon as possible to avoid being blocked by this, I think this is something we can change at any time.

Comment (ethib137) [reply modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: I definitely really want to get this unblocked as well, but I don't think we want to merge it like this. Notice for instance, we are importing CONSTANTS from frontend-js-react-web, but instead of those constants being related to the react, they are constants for the accessibility menu. 

I think it's worth the time to remove this coupling first.

Comment (matuzalemsteles) [reply modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: Yeah, so I think the best solution here would really be to create the module as suggested and we can discuss it later if this really is the best solution or if there are others with Bryce.

Comment (ethib137) [reply modules/apps/accessibility/accessibility-menu-web/npmscripts.config.js]: Yeah, let's do that. Create a module to export the atom, get this merged and then discuss the best solution when Bryce is back.

Thanks.

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
ci:stop

**Comment by matuzalemsteles** [comment]
ci:test:sf

File: modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx
Comment (ethib137) [inline modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx:17]: Do the destructured variables need to be sorted? Is that why SF is failing?

Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx:17]: @ethib137 maybe it is, but when I change the order it causes an SF error too, I think it's related to the number of columns in the line that has 92 the maximum is 80.

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
Waiting on the results triggered today... I'll continue the analysis Tomorrow!

**Comment by eltonccdantas** [comment]
ci:help

**Comment by eltonccdantas** [comment]
ci:test:commerce

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
After further investigation, @matuzalemsteles and I reached to point that the failures are unrelated to this pull request.

[comparison run - commerce](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=2&testrayRunIdB=2371688213&testrayRunIdA=2371302355&statusB=3)

[comparison run - relevant ](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=2&testrayRunIdB=2371775425&testrayRunIdA=2371302355&statusB=3)

**Comment by eltonccdantas** [comment]
ci:forward:force

**Comment by matuzalemsteles** [comment]
ci:forward:force

**Comment by matuzalemsteles** [comment]
ci:forward:force

**Comment by eltonccdantas** [comment]
ci:forward:force

**Comment by eltonccdantas** [comment]
Manually forwarded to https://github.com/brianchandotcom/liferay-portal/pull/137483

**Pull request closed without merging**
